### ISSUE TITLE
Builder from uses Guava immutables internally and can cause IllegalArgumentException

### ISSUE THREAD

**Issue by Randgalt** [immutables/immutables#1194]
The Immutables builder `from()` method is not very useful with maps. If you want to duplicate a value and change a map key to a different value you cannot do it. You get ` java.lang.IllegalArgumentException: Multiple entries with same key`

```java
    @Value.Immutable
    public interface Test
    {
        Map<String, Things> things();
    }

    @Value.Immutable
    public interface Things
    {
        String name();

        int qty();
    }

    public static void main(String[] args)
    {
        Test test1 = ImmutableTest.builder()
                .putThings("one", ImmutableThings.builder().name("a").qty(1).build())
                .putThings("two", ImmutableThings.builder().name("b").qty(2).build())
                .build();
        Test test2 = ImmutableTest.builder().from(test1)
                .putThings("two", ImmutableThings.builder().name("c").qty(3).build())
                .build();
    }
```

**Comment by elucash** [comment]
yep, this is how it works with Guava, have you tried `Style(jdkOnly=true)` to see if standard JDK collections would allow what you want?

**Comment by Randgalt** [comment]
OK - that works. Why then even have the Guava versions? Also, this required two annotations per interface right?

**Comment by elucash** [comment]
> Why then even have the Guava versions?
Guava based implementation was created first, then jdkOnly, also the behavior you would expect can also be non-starter for other folks: some prefer silent overrides, some prefer to be very explicit of merging and consider accidental override an error.

> Also, this required two annotations per interface right?
There are many ways to reduce clutter, like creating a single style and putting it on top level package or using meta-annotation. see https://immutables.github.io/style.html, there are some examples

**Comment by talios** [comment]
Also, originally Guava’s Function/Predicate/Optional and other classes, as
well as existing before the core platforms versions, didn’t extend from
Java’s until later in their release cycle when JDK8 was quite more
prevalent.

I suppose now that they do, it might be less of an issue, and maybe
jdkOnly=true should become the default - but that’s quite a change that
would more than likely break A LOT of peoples code, for little benefit.

Mark

On 8 July 2020 at 6:23:23 PM, Eugene Lukash (notifications@github.com)
wrote:

Guava based implementation was created first, then jdkOnly, also the
behavior you would expect can also be non-starter for other folks: some
prefer silent overrides, some prefer to be very explicit of merging and
consider accidental override an error.

**Comment by elucash** [comment]
yep, the problems with migration path and backward compatibility etc would be too much. It may be ok for some next-gen version for java 9+ (probably 11+), we would use standard immutable collections (`List.of`, `Set.of`), emulate all guava stuff via pluggable encodings and so on, so on. .. if we ever get to it

**Comment by schlosna** [comment]
Would you all be open to a PR to enable Guava use of `buildKeepingLast()`, e.g. via an option like `keepLastValue = true`? There's a bit of weirdness where with default `jdkOnly = false` we'd want to default `keepLastValue = false` to match existing behavior while `jdkOnly = true` is implicitly `keepLastValue = true` today (unless we switch over to JDK 9+ `Map.of`).

It looks like this would require a minimum Guava version bump as immutables [currently depends on Guava 30.0-jre](https://github.com/immutables/immutables/blob/af0cad7169e83042c81c3778223eb0c4e2b72cd7/pom.xml#L120) (released 2020-10-17) since immutables [2.10.0](https://github.com/immutables/immutables/releases/tag/2.10.0) in https://github.com/immutables/immutables/pull/1422 . Guava 31.1-jre which adds the `ImmutableMap.Builder<K,​V>buildKeepingLast()` was released 2022-03-01. 

https://github.com/immutables/immutables/blob/af0cad7169e83042c81c3778223eb0c4e2b72cd7/pom.xml#L120
